boa/src/syntax/lexer/tests.rs

```diff
@@ -264,283 +152,438 @@ fn check_keywords() {
+        TokenKind::Punctuator(Punctuator::Assign),
+        TokenKind::string_literal("hello"),
+        TokenKind::Punctuator(Punctuator::Semicolon),
+    ];
+
+    expect_tokens(&mut lexer, &expected);
 }
 
 #[test]
 fn check_positions() {
     let s = r#"console.log("hello world\u{2764}"); // Test"#;
     // --------123456789
-    let mut lexer = Lexer::new(s);
-    lexer.lex().expect("failed to lex");
+    let mut lexer = Lexer::new(s.as_bytes());
+
     // The first column is 1 (not zero indexed)
-    assert_eq!(lexer.tokens[0].span(), span((1, 1), (1, 7)));
+    assert_eq!(lexer.next().unwrap().unwrap().span(), span((1, 1), (1, 8)));
 
     // Dot Token starts on column 8
-    assert_eq!(lexer.tokens[1].span(), span((1, 8), (1, 8)));
+    assert_eq!(lexer.next().unwrap().unwrap().span(), span((1, 8), (1, 9)));
 
     // Log Token starts on column 9
-    assert_eq!(lexer.tokens[2].span(), span((1, 9), (1, 11)));
+    assert_eq!(lexer.next().unwrap().unwrap().span(), span((1, 9), (1, 12)));
 
     // Open parenthesis token starts on column 12
-    assert_eq!(lexer.tokens[3].span(), span((1, 12), (1, 12)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 12), (1, 13))
+    );
 
     // String token starts on column 13
-    assert_eq!(lexer.tokens[4].span(), span((1, 13), (1, 33)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 13), (1, 34))
+    );
 
     // Close parenthesis token starts on column 34
-    assert_eq!(lexer.tokens[5].span(), span((1, 34), (1, 34)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 34), (1, 35))
+    );
 
     // Semi Colon token starts on column 35
-    assert_eq!(lexer.tokens[6].span(), span((1, 35), (1, 35)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 35), (1, 36))
+    );
 }
 
 #[test]
-#[ignore]
 fn two_divisions_in_expression() {
     let s = "    return a !== 0 || 1 / a === 1 / b;";
-    let mut lexer = Lexer::new(s);
-    lexer.lex().expect("failed to lex");
-    // dbg!(&lexer.tokens);
+    let mut lexer = Lexer::new(s.as_bytes());
+
+    let expected = [
+        TokenKind::Keyword(Keyword::Return),
+        TokenKind::Identifier("a".into()),
+        TokenKind::Punctuator(Punctuator::StrictNotEq),
+        TokenKind::NumericLiteral(Numeric::Integer(0)),
+        TokenKind::Punctuator(Punctuator::BoolOr),
+        TokenKind::NumericLiteral(Numeric::Integer(1)),
+        TokenKind::Punctuator(Punctuator::Div),
+        TokenKind::Identifier("a".into()),
+        TokenKind::Punctuator(Punctuator::StrictEq),
+        TokenKind::NumericLiteral(Numeric::Integer(1)),
+        TokenKind::Punctuator(Punctuator::Div),
+        TokenKind::Identifier("b".into()),
+        TokenKind::Punctuator(Punctuator::Semicolon)
+    ];
 
-    assert_eq!(lexer.tokens[11].span(), span((1, 37), (1, 37)));
+    expect_tokens(&mut lexer, &expected);
 }
```

Comment:
@Razican 

I've re-written this test to show that the exactly the right tokens are returned and like this, the test passes (now that goal symbols have been partially added in). 

Does this test look ok to you? / does it show that the desired result is being produced or should we be checking span's aswell?

Reply:
This is good for me. Still, since we need to specify the goal symbol, this test will need to change, as the lexer will need to be called only from the parser.

Reply:
Can't we sometimes infer the goal symbol? For example there are times when only a certain goal symbol is valid and I've added some of these into the code

Reply:
Maybe we can, yes, do you have an example?

Reply:
`
_ if next_chr.is_digit(10) => {
                let result = NumberLiteral::new(next_chr, strict_mode).lex(&mut self.cursor, start);
                // A regex may not directly follow a NumericLiteral but a DivPunctuator may.
                // Note that the goal cannot be set to InputElementTemplateTail at this point as a TemplateSubstitutionTail would be invalid.
                self.set_goal(InputElement::Div);
                result
            }
            _ if next_chr.is_alphabetic() || next_chr == '$' || next_chr == '_' => {
                let result = Identifier::new(next_chr).lex(&mut self.cursor, start);

                // A regex may not directly follow an Identifier but a DivPunctuator may.
                // Note that the goal cannot be set to InputElementTemplateTail at this point as a TemplateSubstitutionTail would be invalid.
                self.set_goal(InputElement::Div);
                result
            }

`

Reply:
Yep, most of the times this is what we want. There are only two cases where we depend on the goal symbol: The `/` character, and the template tail.